Add certificate validation cache

HaoK opened the pull request:
First stab at https://github.com/dotnet/aspnetcore/issues/12324

Also fixes https://github.com/dotnet/aspnetcore/issues/22037

HaoK committed:
Add certificate validation cache

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <param name="result">the <see cref="AuthenticateResult"/></param>
+        public void Put(ConnectionInfo connection, X509Certificate certificate, AuthenticateResult result)
```
The auth result shouldn't be shared across requests, no? The Principal is mutable and something like ClaimsTransformation could mess it up. Would you have to do a deep clone when putting it into the cache?

HaoK replied:
Yeah I guess we should use the IPrincipal ctor

HaoK replied:
Added a clone to AuthenticateResult/Ticket to do the deep copy

HaoK replied:
Added a clone to AuthenticateResult/Ticket to do the deep copy

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <param name="certificate">The certificate.</param>
+        /// <param name="result">the <see cref="AuthenticateResult"/></param>
+        public void Put(ConnectionInfo connection, X509Certificate certificate, AuthenticateResult result)
+            =>  _cache.Set(ComputeKey(connection, certificate), result, new MemoryCacheEntryOptions().SetSize(1));
```
Needs a (configurable?) timeout. At the moment items will only get removed from the cache if the cache is full.

The cache size should also be configurable. If you have more concurrent connections than the cache size then you'll get thrashing.

HaoK replied:
Added Size and Timeout as properties to the default implementation

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,39 @@
+            =>  _cache.Set(ComputeKey(connection, certificate), result, new MemoryCacheEntryOptions().SetSize(1));
+
+        private string ComputeKey(ConnectionInfo connection, X509Certificate certificate)
+            => $"{certificate.GetCertHashString()}:{connection.Id}";
```
@blowdart is the cert hash string a strong enough identifier? Or is the connectionId needed too?

It would be even more performant if we could cache the cert validation across connections, but that risks collisions.

blowdart replied:
No. It's SHA1. Compute a SHA256 over the RawData property (and maybe make that an extension method that lives in X509CertificateExtensions.cs for neatness?

If SHA256 has collisions we have bigger problems :)

HaoK replied:
We can just use GetCertHashString(HashAlgorithmName.SHA256), we don't need the connection id then right?

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <returns>the <see cref="AuthenticateResult"/></returns>
+        public AuthenticateResult Get(ConnectionInfo connection, X509Certificate certificate)
```
Might as well pass in the HttpContext and let the implementation decide what the key is composed of. You'll need that anyways if you want to implement one based on kestrel's ConnectinItems Feature.

Tratcher reviewed (commented):
That's the right idea, though caching has a lot of nits to deal with.

HaoK commented on src/Security/Authentication/Certificate/src/CertificateAuthenticationHandler.cs:
```diff
@@ -60,57 +63,15 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
-                // any other validations.
-                if (clientCertificate.IsSelfSigned() &&
-                    !Options.AllowedCertificateTypes.HasFlag(CertificateTypes.SelfSigned))
+                var cacheHit = _cache.Get(Context.Connection, clientCertificate);
```
Maybe we can just make it opt in via the Options then, that would take care of the ctor break as well.

HaoK replied:
Updated to just use the interface and DI for opt in, InitializeHandler service locates

HaoK committed:
CR feedback

CLeanup

Apply suggestions from code review

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

Update CertificateValidationCache.cs

Clone a little deeper

Clone cleanup

Update comment

Update refs

Update src/Security/Authentication/Certificate/src/CertificateValidationCache.cs

Co-authored-by: James Newton-King <james@newtonking.com>

Fix comment

Whitespace

Enable caching in cert sample

Fix ref

Update CertificateAuthenticationHandler.cs

Update LoggingExtensions.cs

Fix

Tratcher commented on src/Http/Authentication.Abstractions/src/AuthenticateResult.cs:
```diff
@@ -46,6 +46,28 @@ public class AuthenticateResult
+                return Success(Ticket.Clone());
+            }
+            // This shouldn't happen
+            return this;
```
```suggestion
            throw new NotImplementedException();
```

Tratcher commented on src/Http/Authentication.Abstractions/src/AuthenticationTicket.cs:
```diff
@@ -52,5 +52,13 @@ public AuthenticationTicket(ClaimsPrincipal principal, string authenticationSche
+        /// </summary>
+        /// <returns>A copy of the ticket</returns>
+        public AuthenticationTicket Clone()
+            => new AuthenticationTicket(new ClaimsPrincipal(Principal), new AuthenticationProperties(Properties.Items, Properties.Parameters), AuthenticationScheme);
```
These are still shallow clones for Principal and Properties. It's still the same mutable Identity instances and property collections.

HaoK replied:
I added a deeper clone for properties and used Principal.Clone which still results in cloning the identities, is it reasonable for us to just iterate over each identity in the principal and create a new claimsprincipal and basically do a 'claims' snapshot clone of the principal for the cache?

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,57 @@
+        public void Put(HttpContext context, X509Certificate2 certificate, AuthenticateResult result)
+            =>  _cache.Set(ComputeKey(context, certificate), result, new MemoryCacheEntryOptions().SetSize(1).SetSlidingExpiration(CacheEntryExpiration));
+
+        private string ComputeKey(HttpContext context, X509Certificate2 certificate)
```
```suggestion
        private string ComputeKey(X509Certificate2 certificate)
```

Tratcher reviewed (commented):
Update the sample project to show enabling and configuring the cache?

We don't want to put a Kestrel specific implementation in this package for dependency reasons (I think?), but we should at least try to build that version in the sample.

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <summary>
+        /// The expiration that should be used for entries in the MemoryCache, defaults to 30 minutes.
+        /// </summary>
+        public TimeSpan CacheEntryExpiration { get; set; } = TimeSpan.FromMinutes(30);
```
```suggestion
        public TimeSpan CacheEntryExpiration { get; set; } = TimeSpan.FromMinutes(2);
```
2 minutes is the average idle connection timeout.

Tratcher reviewed (commented):
ClaimsPrincipal.Clone does the same shallow copy. ClaimsIdentity defines its own Clone that does do a copy all the way down to primitives.

HaoK commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <param name="certificate">The certificate.</param>
+        /// <returns>the <see cref="AuthenticateResult"/></returns>
+        public AuthenticateResult Get(HttpContext context, X509Certificate2 certificate)
+            => _cache.Get(ComputeKey(certificate)) as AuthenticateResult;
```
Yeah not yet, will do that now that Clone should be working with the new iteration

HaoK commented:
It seems a little weird to me that ClaimsPrincipal.Clone isn't doing what we want, and are doing our own thing to call clone on ClaimsPrincipal.Identities, but I guess we can just document this in the doc comments

JamesNK commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,60 @@
+
+
+        /// <summary>
+        /// The expiration that should be used for entries in the MemoryCache, defaults to 30 minutes.
```
FYI, comment of "30 minutes" doesn't match 2 minutes in code.

JamesNK commented:
The semantics of clone have been a garbage fire since .NET Framework 1.0 😄

JamesNK commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,60 @@
+        public AuthenticateResult Get(HttpContext context, X509Certificate2 certificate)
+        {
+            var result = _cache.Get(ComputeKey(certificate)) as AuthenticateResult;
+            return result ?? result.Clone();
```
```suggestion
            return result?.Clone();
```

JamesNK commented:
What is the plan to perf test this?

HaoK commented:
We don't have any perf for security generally, do you have any suggestions on how we could see if this helps @JamesNK ?

davidfowl commented:
@HaoK er need a performance test, work with @sebastienros to get one setup and integrated. We shouldn’t be making huge perf changes like this without before and after numbers

blowdart commented:
Yup, need perf numbers before signoff

Tratcher commented on src/Http/Authentication.Abstractions/src/AuthenticationProperties.cs:
```diff
@@ -45,6 +45,13 @@ public AuthenticationProperties(IDictionary<string, string> items, IDictionary<s
+        /// </summary>
+        /// <returns>A copy.</returns>
+        public AuthenticationProperties Clone()
+            => new AuthenticationProperties(new Dictionary<string, string>(Items), new Dictionary<string, object>(Parameters));
```
Should still set the comparer on the new dictionaries.

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,59 @@
+        /// <returns>the <see cref="AuthenticateResult"/></returns>
+        public AuthenticateResult Get(HttpContext context, X509Certificate2 certificate)
+        {
+            var result = _cache.Get(ComputeKey(certificate)) as AuthenticateResult;
```
There's a generic extension isn't there?
```suggestion
            var result = _cache.Get<AuthenticateResult>(ComputeKey(certificate));
```
https://github.com/dotnet/runtime/blob/e5516bb2b188e56b47a758c9cbf0a7f1adf90dd0/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs#L19

Tratcher reviewed (commented):
Looks pretty good. Now for perf tests.

JamesNK commented:
FYI,

I updated the gRPC server and client to support client certificates: https://github.com/grpc/grpc-dotnet/pull/891/commits/162cf731202534ffe8d5728ed6306f72170c4162

You can run them from https://github.com/JamesNK/grpc-dotnet/tree/jamesnk/benchmarch-grpcclient branch.

Server:
`dotnet run -p .\perf\benchmarkapps\GrpcAspNetCoreServer\ -c Release --protocol h2 --enableCertAuth true`

Client:
`dotnet run -c Release -p .\perf\benchmarkapps\GrpcClient\ -- -u http://localhost:5000 -c 100 -s unary -p h2 --grpcClientType grpcnetclient --clientCertificate true`

Numbers I saw (running on 3.1.3):
* Without cert auth: 43,451
* With cert auth: 24,274

HaoK committed:
Sugar

Tratcher commented on src/Security/Authentication/Certificate/samples/Certificate.Sample/Startup.cs:
```diff
@@ -37,6 +37,8 @@ public void ConfigureServices(IServiceCollection services)
                 });
 
+            // Adding a ICertificateValidationCache will result in certificate auth caching the results, the default implementation uses a memory cache
+            services.AddSingleton<ICertificateValidationCache, CertificateValidationCache>();
```
Should we add an extension for this on the auth builder for discovery?

HaoK replied:
like service.AddCertificate().AddDefaultCache()?

Tratcher replied:
Yeah,
```
services.AddAuthentication(CertificateAuthenticationDefaults.AuthenticationScheme)
                .AddCertificate(...)
                .AddCertificateCache(options => {})
```

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public CertificateValidationCache()
+        {
+             _cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = CacheSize });
```
Aren't you reading CacheSize before anyone has a chance to change it? Should this be using `IOptions<CertificateValidationCache>`?

HaoK replied:
Well I expected it to be more configured by derivation, MyCache : CertificateValidationCache and they change the properties

HaoK replied:
Do we really need a top level options for this?

Tratcher replied:
> Well I expected it to be more configured by derivation, MyCache : CertificateValidationCache and they change the properties

That's not a pattern we use anywhere else is it?

> Do we really need a top level options for this?

That's the standard pattern.

HaoK replied:
Well I mostly meant that I didn't think tweaking the cache slightly was going to be the main idea, I thought perhaps mistakenly that we would expect people to just implement their own cache and drop it in mostly.  But if we are going to add first class sugar, it makes sense to expose options

HaoK committed:
Add options sugar

HaoK commented:
Alright moved the 2 settings into their own separate options class and added the AddCertificateCache sugar to AuthBuilder

HaoK committed:
Update AuthenticationProperties.cs

HaoK commented:
Marking as ready for review, been working with @sebastienros and since this is new API, and we've discovered that it would be much easier for us to get this running on the official performance runs if we get this in and measure it the real way, rather than building self contained apps with patched dlls.

Tratcher commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <param name="certificate">The certificate.</param>
+        /// <param name="result">the <see cref="AuthenticateResult"/></param>
+        public void Put(HttpContext context, X509Certificate2 certificate, AuthenticateResult result)
+            => _cache.Set(ComputeKey(certificate), result.Clone(), new MemoryCacheEntryOptions().SetSize(1).SetSlidingExpiration(_options.CacheEntryExpiration));
```
```suggestion
            => _cache.Set(ComputeKey(certificate), result.Clone(), new MemoryCacheEntryOptions()
                .SetSize(1).SetSlidingExpiration(_options.CacheEntryExpiration).SetAbsoluteExpiration(certificate.NotAfter));
```

HaoK committed:
Apply suggestions from code review

Co-authored-by: Chris Ross <Tratcher@Outlook.com>

JamesNK commented:
What are the benchmark results after this change?

HaoK commented:
We had trouble getting the perf rig setup to measure the numbers due to new surface API numbers, so we are merging this so we can flow the new API to the performance infrastructure and run the tests

Tratcher commented:
Sorry, merge conflict.

HaoK committed:
Merge branch 'master' into haok/cert

JamesNK commented on src/Security/Authentication/Certificate/src/CertificateValidationCache.cs:
```diff
@@ -0,0 +1,48 @@
+                .SetSize(1).SetSlidingExpiration(_options.CacheEntryExpiration).SetAbsoluteExpiration(certificate.NotAfter));
+
+        private string ComputeKey(X509Certificate2 certificate)
+            => $"{certificate.GetCertHashString(HashAlgorithmName.SHA256)}";
```
```suggestion
            => certificate.GetCertHashString(HashAlgorithmName.SHA256);
```

HaoK committed:
Update src/Security/Authentication/Certificate/src/CertificateValidationCache.cs

Co-authored-by: James Newton-King <james@newtonking.com>

HaoK commented:
@sebastienros do we need to do anything special to get this to flow to your performance tools, or do we just wait for the normal process to happen?

sebastienros commented:
It's nothing special to the benchmarks. The server is using nightly sdks, and they fail to build the app as this service is unknown by the sdk. It's the same issue whenever we use new aspnet apis. We just have to wait for the sdk to ingest a new aspnet version. At that point our apps can build on the servers.

I don't think this should take long now.
